django/contrib/postgres/fields/ranges.py

```diff
@@ -132,6 +133,44 @@ def db_type(self, connection):
 RangeField.register_lookup(lookups.Overlap)
 
 
+class DateTimeRangeContains(models.Lookup):
+    lookup_name = 'contains'
+
+    def as_sql(self, compiler, connection):
+        # Transform rhs value for db lookup
+        value = self.rhs
+        if isinstance(value, datetime.datetime):
+            output_field = models.DateTimeField()
+        elif isinstance(value, datetime.date):
+            output_field = models.DateField()
+        else:
+            output_field = None
+
+        if output_field:
+            value = models.Value(value, output_field=output_field)
+            self.rhs = value.resolve_expression(compiler.query)
```

Comment:
It seems a bit odd to be that we assign a new value to `self.rhs` at SQL compilation time. Isn't it something that can be done at initialization time?

```python
def __init__(self, lhs, rhs):
    if isinstance(rhs, datetime.date):
        output_field = models.DateTimeField() if isintance(rhs, datetime.datetime) else models.DateField()
        rhs = models.Value(rhs, output_field=output_field)
    super(self, DateTimeRangeContains).__init__(lhs, rhs)
```